### PR TITLE
use txo list for wallet page:

### PR DESCRIPTION
**Pull request by jessopb**
paginated
enable revoking
filtering
separate txo list from transaction list

### CODE REVIEW

File: ui/constants/txo_list.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+export const ACTIVE = 'active'; // spent, active, all
```
Comment (seanyesmunt) [inline ui/constants/txo_list.js:1]: should these be in lbry-redux?

File: ui/component/txoList/view.jsx [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+      title={__(`Transactions`)}
+      titleActions={
+        <div className="card__actions--inline">
+          <Button button="secondary" label={__('Refresh')} onClick={() => fetchTransactionPage(params)} />
+          {/* @if TARGET='app' */}
+          <Button button="primary" navigate={`/$/${PAGES.TRANSACTIONS}`} label={__('Full History')} />
+          {/* @endif */}
+        </div>
+      }
+      headerActions={
```
Comment (seanyesmunt) [inline ui/component/txoList/view.jsx:181]: Instead of creating this I think this whole section should just be moved into the `body`. You can wrap it in `.section--padded` if it needs spacing due to the weird tableeness

File: ui/component/txoList/view.jsx [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+                      <option key={stringV} value={stringV}>
+                        {stringV && __(toCapitalCase(stringV))}
+                      </option>
+                    );
+                  })}
+                </FormField>
+              </div>
+              {/* show the subtypes drop down */}
+              {(type === TXO.SENT || type === TXO.RECEIVED) && (
+                <div>
```
Comment (seanyesmunt) [inline ui/component/txoList/view.jsx:206]: There are a few extra divs in this component

File: ui/component/txoList/view.jsx [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    pageSize,
+    active,
+    type,
+    subtype,
+  };
+
+  // build new json params
+  const params = {};
+
+  if (currentUrlParams.type) {
```
Comment (seanyesmunt) [inline ui/component/txoList/view.jsx:53]: 😳  I commend you for figuring this logic out.

File: ui/component/txoList/view.jsx [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+import React, { useEffect } from 'react';
+import { withRouter } from 'react-router';
+
+import { TX_LIST } from 'lbry-redux';
+import * as PAGES from 'constants/pages';
+import * as TXO from 'constants/txo_list';
+
+import TransactionListTable from 'component/transactionListTable';
+import Paginate from 'component/common/paginate';
+// import FileExporter from '../common/file-exporter'; // reimplement as modal
```
Comment (seanyesmunt) [inline ui/component/txoList/view.jsx:11]: if it's a lot of work, don't worry about adding this back for the first implementation. We can add it back once the dust settles

File: ui/component/transactionListTable/view.jsx [hunk 6]
```diff
@@ -2,37 +2,32 @@
...
   }
 
   return (
     <React.Fragment>
-      {!loading && !transactionList.length && (
-        <h2 className="main--empty empty">{emptyMessage || __('No transactions.')}</h2>
-      )}
-      {!!transactionList.length && (
+      {!loading && !list.length && <h2 className="main--empty empty">{emptyMessage || __('No transactions.')}</h2>}
+      {loading && <h2 className="main--empty empty">{__('Loading...')}</h2>}
```
Comment (seanyesmunt) [inline ui/component/transactionListTable/view.jsx:29]: Replace `loading...` with `<Spinner delayed />`

File: ui/component/transactionListTable/view.jsx [hunk 7]
```diff
@@ -2,37 +2,32 @@
...
-    const outpoint = `${txid}:${nout}`;
-    const { mySupports, myClaims } = props;
-
-    return !!mySupports[outpoint] || myClaims.has(outpoint);
-  }
-
-  function revokeClaim(txid: string, nout: number) {
-    props.openModal(MODALS.CONFIRM_CLAIM_REVOKE, { txid, nout });
+  const { emptyMessage, rewards, loading, txos, transactions } = props;
+  const REVOCABLE = ['channel', 'stream', 'repost', 'support', 'claim'];
```
Comment (seanyesmunt) [inline ui/component/transactionListTable/view.jsx:20]: REVOCABLE_TYPES

File: ui/component/txoList/view.jsx [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+  useEffect(() => {
+    if (paramsString && fetchTransactionPage) {
+      const params = JSON.parse(paramsString);
+      fetchTransactionPage(params);
+    }
+  }, [paramsString, fetchTransactionPage]);
+
+  return (
+    <Card
```
Comment (seanyesmunt) [inline ui/component/txoList/view.jsx:171]: This should use the same style as here
https://github.com/lbryio/lbry-desktop/blob/master/ui/component/transactionList/view.jsx#L55

If we use a <Card /> we can't get the table to extend the full width of the background

File: ui/component/transactionListTable/internal/transaction-list-item.jsx [hunk 9]
```diff
@@ -80,7 +59,7 @@ class TransactionListItem extends React.PureComponent<Props> {
           )}
         </td>
         <td className="table__item--actionable">
-          <span>{this.capitalize(type)}</span> {isRevokeable && this.getLink(type)}
+          <span>{this.capitalize(type)}</span>{' '}
```
Comment (seanyesmunt) [inline ui/component/transactionListTable/internal/transaction-list-item.jsx:62]: extra space

File: ui/component/common/card.jsx [hunk 10]
```diff
@@ -46,18 +50,22 @@ export default function Card(props: Props) {
...
-              <Button
-                button={'alt'}
-                aria-label={__('More')}
-                icon={expanded ? toCapitalCase(ICONS.SUBTRACT) : toCapitalCase(ICONS.ADD)}
-                onClick={() => setExpanded(!expanded)}
-              />
-            </div>
-          )}
+          <div>
+            {titleActions && <div className="section--padded">{titleActions}</div>}
```
Comment (seanyesmunt) [inline ui/component/common/card.jsx:54]: I'm not that big of a fan of `titleActions`, but I think it's fine to keep. I don't think `headerActions` should be added though.

Comment (kauffj) [reply ui/component/common/card.jsx:54]: not section--padded though, just `.card__header-actions` (or whatever) (even if duplicating CSS)

File: ui/component/transactionListTable/internal/txo-list-item.jsx [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                <DateTime date={date} show={DateTime.SHOW_TIME} />
+              </div>
+            </div>
+          ) : (
+            <span className="empty">{__('Pending')}</span>
+          )}
+        </td>
+        <td className="table__item--actionable">
+          <span>
+            {(isTip && this.capitalize('tip')) || (valueType && this.capitalize(valueType)) || this.capitalize(type)}
```
Comment (kauffj) [inline ui/component/transactionListTable/internal/txo-list-item.jsx:112]: i18n

File: ui/component/transactionListTable/view.jsx [hunk 12]
```diff
@@ -45,15 +40,24 @@ function TransactionListTable(props: Props) {
...
-              {transactionList.map(t => (
-                <TransactionListItem
-                  key={`${t.txid}:${t.nout}`}
-                  transaction={t}
-                  reward={rewards && rewards[t.txid]}
-                  isRevokeable={isRevokeable(t.txid, t.nout)}
-                  revokeClaim={revokeClaim}
-                />
-              ))}
+              {transactions &&
```
Comment (kauffj) [inline ui/component/transactionListTable/view.jsx:43]: we're moving to two entirely separate ways of rendering transactions? this seems like a step backwards - why is this necessary?

File: ui/constants/txo_list.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+export const PAYMENT = 'payment';
+export const DROPDOWN_SUBTYPES = [ALL, TIP, PURCHASE, PAYMENT];
+
+// rpc params
+export const TX_TYPE = 'type'; // = other, stream, repost, channel, support, purchase
+export const IS_SPENT = 'is_spent';
+export const IS_NOT_SPENT = 'is_not_spent';
+export const IS_MY_INPUT = 'is_my_input';
+export const IS_MY_OUTPUT = 'is_my_output';
+export const IS_NOT_MY_INPUT = 'is_not_my_input';
```
Comment (kauffj) [inline ui/constants/txo_list.js:27]: why are these NOT parameters needed? isn't `IS_MY_INPUT=true|false|null` sufficient?

File: ui/component/txoList/view.jsx [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+      params[TXO.IS_NOT_SPENT] = true;
+    }
+  }
+
+  if (currentUrlParams.page) params[TXO.PAGE] = Number(page);
+  if (currentUrlParams.pageSize) params[TXO.PAGE_SIZE] = Number(pageSize);
+
+  function handleChange(delta: Delta) {
+    const url = updateUrl(delta);
+    history.push(url);
```
Comment (kauffj) [inline ui/component/txoList/view.jsx:102]: hmmm... replace or push? Do I need/want to paginate through every filter change? (unsure on this)

File: ui/component/txoList/view.jsx [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+  }
+
+  function updateUrl(delta: Delta) {
```
Comment (kauffj) [inline ui/component/txoList/view.jsx:105]: should any of this logic be moved into selectors? also looks like some repeated code around page size, possibly more

File: ui/component/txoList/view.jsx [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+  }, [paramsString, fetchTransactionPage]);
+
+  return (
+    <Card
+      title={__(`Transactions`)}
+      titleActions={
+        <div className="card__actions--inline">
+          <Button button="secondary" label={__('Refresh')} onClick={() => fetchTransactionPage(params)} />
+          {/* @if TARGET='app' */}
+          <Button button="primary" navigate={`/$/${PAGES.TRANSACTIONS}`} label={__('Full History')} />
```
Comment (kauffj) [inline ui/component/txoList/view.jsx:177]: why is this app only?

Comment (jessopb) [reply ui/component/txoList/view.jsx:177]: Relies on full transaction list that blows up sdk memory.

File: ui/component/txoList/view.jsx (hunk 3 above)
Comment (jessopb) [reply ui/component/txoList/view.jsx:206]: I hit some nth of type top and side padding in FormField unless I wrapped in div.

File: ui/component/txoList/view.jsx (hunk 16 above)
Comment (jessopb) [reply ui/component/txoList/view.jsx:177]: There was a conversation about the export feature and the lack of fees info from txo_list that resulted in trying to save that component, at least for desktop app.

File: ui/component/transactionListTable/view.jsx (hunk 12 above)
Comment (jessopb) [reply ui/component/transactionListTable/view.jsx]: deleted

File: ui/constants/txo_list.js (hunk 13 above)
Comment (jessopb) [reply ui/constants/txo_list.js:27]: Those are the actual sdk flags.

File: ui/constants/txo_list.js (hunk 1 above)
Comment (jessopb) [reply ui/constants/txo_list.js]: done

File: ui/component/txoList/view.jsx (hunk 15 above)
Comment (jessopb) [reply ui/component/txoList/view.jsx:102]: I can move all the logic in the redux action and pass a json string instead... Let's confirm this is a good idea.

Comment (jessopb) [reply ui/component/txoList/view.jsx:99]: This can/will be done later.

File: ui/component/txoList/view.jsx (hunk 14 above)
Comment (jessopb) [reply ui/component/txoList/view.jsx:96]: hm, likely replace, as navigating back to page 20 after deleting a bunch of stuff is dumb.

File: ui/component/transactionListTable/internal/txo-list-item.jsx [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+        icon={ICONS.DELETE}
+        onClick={this.abandonClaim}
+        title={__(abandonTitle)}
+      />
+    );
+  }
+
+  abandonClaim() {
+    this.props.revokeClaim(this.props.txo, st => this.setState({ abandonState: st }));
+    // this.setState({ disabled: true }); // just flag the item disabled
```
Comment (seanyesmunt) [inline ui/component/transactionListTable/internal/txo-list-item.jsx:62]: comment

File: ui/component/transactionListTable/internal/txo-list-item.jsx [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+        button="secondary"
+        icon={ICONS.DELETE}
+        onClick={this.abandonClaim}
+        title={__(abandonTitle)}
+      />
+    );
+  }
+
+  abandonClaim() {
+    this.props.revokeClaim(this.props.txo, st => this.setState({ abandonState: st }));
```
Comment (seanyesmunt) [inline ui/component/transactionListTable/internal/txo-list-item.jsx:61]: I would rename `st` to something more meaningful. You could just call it `abandonState`

File: ui/component/transactionListTable/internal/txo-list-item.jsx [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+    // this.setState({ disabled: true }); // just flag the item disabled
+  }
+
+  capitalize = (string: string) => string && string.charAt(0).toUpperCase() + string.slice(1);
```
Comment (seanyesmunt) [inline ui/component/transactionListTable/internal/txo-list-item.jsx:65]: Can you move this to `/util`. there is another file that uses the same thing

File: ui/component/transactionListTable/internal/txo-list-item.jsx [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+
+    const {
+      amount,
+      claim_id: claimId,
+      normalized_name: txoListName,
+      timestamp,
+      txid,
+      type,
+      value_type: valueType,
+      is_my_input: isMyInput, // no transaction
```
Comment (seanyesmunt) [inline ui/component/transactionListTable/internal/txo-list-item.jsx:78]: what does `no transaction` mean?

File: ui/component/txoList/view.jsx [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+                <fieldset-section>
+                  <label>Status</label>
+                  <div className={'txo__radios'}>
+                    <Button
+                      button="alt"
+                      onClick={e => handleChange({ dkey: TXO.ACTIVE, value: 'active' })}
+                      className={classnames(`button-toggle`, {
+                        'button-toggle--active': active === TXO.ACTIVE,
+                      })}
+                      label={__(toCapitalCase('active'))}
```
Comment (seanyesmunt) [inline ui/component/txoList/view.jsx:227]: Why not just rename to `__("Active")`?

File: ui/component/txoList/view.jsx [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+                      className={classnames(`button-toggle`, {
+                        'button-toggle--active': active === 'all',
+                      })}
+                      label={__(toCapitalCase('all'))}
+                    />
+                  </div>
+                </fieldset-section>
+              </div>
+            </div>
+            {/* show active/spent */}
```
Comment (seanyesmunt) [inline ui/component/txoList/view.jsx:249]: comment

File: ui/component/common/card.jsx [hunk 23]
```diff
@@ -48,16 +50,19 @@ export default function Card(props: Props) {
...
-            </div>
-          )}
+          <div>
+            {titleActions && <div className="card__title-actions">{titleActions}</div>}
+            {expandable && (
+              <div className="card__title-actions">
+                <Button
+                  button={'alt'}
+                  aria-label={__('More')}
+                  icon={expanded ? toCapitalCase(ICONS.SUBTRACT) : toCapitalCase(ICONS.ADD)}
```
[the remark this answers is not in the document]
Comment (jessopb) [reply ui/component/common/card.jsx:60]: facepalm, I see why that's there, it was icon={toCapitalCase(type)} to generate the ICON name.
